### PR TITLE
Remove pkg_resource and replace for importlib.metadata in autoinstrumentation

### PR DESCRIPTION
**Pull request by Rodrigo-Novas**
# Description

Pkgresource dependency is [deprecated](https://setuptools.pypa.io/en/latest/pkg_resources.html#package-discovery-and-resource-access-using-pkg-resources), on this pr i replaced pkg_resource for importlib.metadata.

Fixes https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2180

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Run with tox the unit tests

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by ocelotl** [comment]
@Rodrigo-Novas please sign the CLA :v:

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (mariojonke) [inline opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:22]: you might want to remove the `setuptools` dependency in the [pyproject.toml](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/opentelemetry-instrumentation/pyproject.toml#L28) file since it shouldn't be required anymore.

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/bootstrap.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/bootstrap.py:89]: It would seem that this snnipet is correct, I don't see any improvements.

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:51]: Okay, maybe we can leave this for the future, there is also the possibility of using
`entry_points(group="opentelemetry_distro") ` but in older versions like 3.9 and 3.8 the group parameter does not exist, so I don't think it is a valid option

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:69]: that snipet seems fine to me

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (mariojonke) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:51]: the `opentelmetry-api` is also making use of `entry_points` and there it was solved by taking a dependency on the [importlib_metdata](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/pyproject.toml#L30) (imports are happening [here|https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/util/_importlib_metadata.py]). This might have some side effects though, see: https://github.com/open-telemetry/opentelemetry-python/issues/3234.

Another option would be to implement/use a unified version of the `entry_points` API that works with the python builtin `importlib.metadata` module. E.g. the code below should work with all the current python versions:
```python
from sys import version_info

if version_info.minor in (8, 9):
    # in python 3.8 and 3.9 'entry_points' returns a dict. For a unified API
    #   the 'EntryPoints' class is (partly) copied from a later release
    #   https://github.com/python/importlib_metadata/blob/v6.1.0/importlib_metadata/__init__.py#L277
    #

    import itertools
    from importlib.metadata import EntryPoint
    from importlib.metadata import entry_points as _importlib_entry_points
    from importlib.metadata import version

    def _matches_ep(entry_point: EntryPoint, **params) -> bool:
        return (
            params.get("name", entry_point.name) == entry_point.name
            and params.get("group", entry_point.group) == entry_point.group
            and params.get("value", entry_point.value) == entry_point.value
        )

    class EntryPoints(tuple):
        __slots__ = ()

        def __getitem__(self, name):  # -> EntryPoint:
            try:
                return next(iter(self.select(name=name)))
            except StopIteration:
                raise KeyError(name)

        def select(self, **params):
            return EntryPoints(
                ept for ept in self if _matches_ep(ept, **params)
            )

        @property
        def names(self):
            return {ep.name for ep in self}

        @property
        def groups(self):
            return {ep.group for ep in self}

    def entry_points(**params) -> EntryPoints:
        eps = itertools.chain.from_iterable(_importlib_entry_points().values())
        return EntryPoints(eps).select(**params)

elif version_info.minor in (10, 11):
    # in 3.10 and 3.11 when called without parameters 'entry_points' returns
    #   a dict-like (SelectableGroups) object. Called with parameters a list-like
    #   (EntryPoints) object is returned. For a unified API the SelectableGroups
    #   return value is avoided and always the 'EntryPoints' object is returned.

    from importlib.metadata import version  # pylint:disable=no-name-in-module
    from importlib.metadata import (  # pylint:disable=no-name-in-module
        EntryPoint,
        EntryPoints,
        SelectableGroups,
    )
    from importlib.metadata import (
        entry_points as _importlib_entry_points,  # pylint:disable=no-name-in-module
    )

    def entry_points(**params) -> EntryPoints:
        eps = _importlib_entry_points(**params)
        if isinstance(eps, SelectableGroups):
            return getattr(eps, "_all")
        return eps

else:
    from importlib.metadata import (  # pylint:disable=no-name-in-module
        EntryPoint,
        EntryPoints,
        entry_points,
        version,
    )

__all__ = ["entry_points", "version", "EntryPoint", "EntryPoints"]
```

Which of the two approaches is the better, i don't know.

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/bootstrap.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/bootstrap.py:20]: mmm i think maybe the 18.0, because that version solved the https://github.com/pypa/packaging/issues/94 issue related with the requirements.

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:51]: Using a constraint on a version of importlib-metadata would ensure that we could use groups and thus avoid having to use .get on Entry_Points and that's a good thing. On the other hand, I see that we would be restricting a version which could cause compatibility problems.
There is an interesting [post](https://iscinumpy.dev/post/bound-version-constraints/#when-is-it-okay-to-set-an-upper-limit) that report this.

The second approach is fine but I think it would make it difficult for us to maintain the code in the future, every time Python releases a new version we would have to update it and perhaps have to implement a new method if the imp changes. I think it would add an unnecessary workload to us in the future. Additionally, we would be repeating already made code.


If I have to take sides with something, it is to use .get() for the moment.
**But in case I cannot use the .get() approach I would use the version constraint solution**

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/dependencies.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/dependencies.py:27]: I agree it is a good approach

File: opentelemetry-instrumentation/pyproject.toml
Comment (mariojonke) [inline opentelemetry-instrumentation/pyproject.toml:29]: the upper bound probably should be `< 8.0`, see https://github.com/open-telemetry/opentelemetry-python/issues/3570 (or maybe no boundary at all, but not sure about that)

**Review by mariojonke — APPROVED** [review]

**Review by mariojonke — APPROVED** [review]

File: opentelemetry-instrumentation/pyproject.toml
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/pyproject.toml]: Although I don't like version constraints, in this case I would leave it because of cpython's reputation for changing its interface 🤣

**Comment by WillDaSilva** [comment]
I believe this would also fix:
- https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1970

**Comment by psi29a** [comment]
Hope this gets picked back up! Great work so far!

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (mlasch) [inline opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:22]: You are adding `importlib-metadata` to the dependencies, but this import is from the Python standard library. There is a discussion here: https://github.com/open-telemetry/opentelemetry-python/issues/3234 about using  `importlib-metadata` and its implications. Note my comment at the end.

Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:22]: Hello, sorry for the delay. 
If I understand correctly, you mean having a standard library with its own constraint within Opentelemetry would be wrong. I also understand that the change of importlib-metadata version could be made later if an error is reported.

https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2181#discussion_r1504854087

**Comment by pablo-sumup** [comment]
Hey there! Could you prioritize the code review for this PR? It's for migrating to Python 3.12. Thanks a bunch! :D

**Comment by theunkn0wn1** [comment]
Any ETA on this?
the package is broken in modern python versions due to undeclared `pkg_resources` usage.

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (theunkn0wn1) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:22]: IMHO its fine to depend on `importlib-metadata`, as its the python-version independent backport of the stdlib feature.
it will be easier to develop against.

**Comment by lzchen** [comment]
@Rodrigo-Novas 

Are you still working on this? If so we can prioritize getting this reviewed and merged. It is difficult to know what the current status is so please close the conversations that you think have been resolved and note down the open topics that still need to be discussed. Feel free to also join the Python SIG if you want to bring this issue up to the community specifically.

**Comment by michaelgmiller1** [comment]
We are also getting hit with this on every unit test run. It would be great to get a fix in.

**Comment by Rodrigo-Novas** [comment]
> @Rodrigo-Novas
> 
> Are you still working on this? If so we can prioritize getting this reviewed and merged. It is difficult to know what the current status is so please close the conversations that you think have been resolved and note down the open topics that still need to be discussed. Feel free to also join the Python SIG if you want to bring this issue up to the community specifically.

@lzchen 

Its ok for me, you can review it!

**Comment by Rodrigo-Novas** [comment]
> We are also getting hit with this on every unit test run. It would be great to get a fix in.

i will try to fix it

File: CHANGELOG.md
Comment (lzchen) [inline CHANGELOG.md:21]:
```suggestion
  ([#2181](https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2181))
```

File: opentelemetry-instrumentation/pyproject.toml
Comment (Arnatious) [inline opentelemetry-instrumentation/pyproject.toml:30]: We already vendor importlib-metadata in opentelemetry-api (https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/util/_importlib_metadata.py) - should we reuse this rather than re-introduce the dependency?

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (lzchen) [inline opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:50]: I don't think this method of selection is supported for Python 3.8 and 3.9 -> https://docs.python.org/3/library/importlib.metadata.html#entry-points

Comment (Arnatious) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:50]: Related to my earlier comment https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2181#discussion_r1737100608 - if we use the importlib_metadata already taken as a dependency in opentelemetry-api, it's version >=6.4.0 which supports this syntax.

Comment (Arnatious) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:50]:
```
>>> for entry_point in importlib_metadata.entry_points(group="console_scripts"):
...   print(entry_point)
... 
EntryPoint(name='pip', value='pip._internal.cli.main:main', group='console_scripts')
EntryPoint(name='pip3', value='pip._internal.cli.main:main', group='console_scripts')
EntryPoint(name='pip3.9', value='pip._internal.cli.main:main', group='console_scripts')
>>> importlib_metadata.version('importlib_metadata')
'6.4.0'
```

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (lzchen) [inline opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:45]: Is there a more elegant way to obtain the `dist` for a specific `entrypoint` rather than generating and storing info on all installed `dist` s?

Comment (Arnatious) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:45]: I don't believe so - importlib doesn't do any sort of caching like pkg_resources did, so this is the most performant way to do this lookup.

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (Arnatious) [inline opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:35]: Consider making this a `functools.cached_property`?

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:50]: Maybe I'm misunderstanding but we discuss about that here https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2181#discussion_r1494303135.

And here https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2181#discussion_r1504854087 about the version constraint.

> Using a constraint on a version of importlib-metadata would ensure that we could use groups and thus avoid having to use .get() on Entry_Points and that's a good thing. On the other hand, I suppose that we would be restricting a version which could cause compatibility problems.
> There is an interesting [post](https://iscinumpy.dev/post/bound-version-constraints/#when-is-it-okay-to-set-an-upper-limit) that report this.
> 
> The second approach is fine but I think it would make it difficult for us to maintain the code in the future, every time Python releases a new version we would have to update it and perhaps have to implement a new method if the imp changes. I think it would add an unnecessary workload to us in the future. Additionally, we would be repeating already made code.
> 
> If I have to take sides with something, it is to use .get() at the moment.
> But in case we cannot use the .get() approach I would use the version constraint solution


@xrmx @Arnatious

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:35]: Something like this?:

```
    @cached_property
    def dist_for(self, entry_point: EntryPoint):
        dist = getattr(entry_point, "dist", None)
        if dist:
            return dist

        if self._mapping is None:
            self._mapping = {
                self._key_for(ep): dist
                for ep in dist.entry_points
                for dist in distributions()
            }

        return self._mapping.get(self._key_for(entry_point))
```



This would always return the same dist, right?

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:101]: entry_point.dist used in [line 77](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2181/files#diff-6ecd6cbdf97812bf55619474db901385d54a6a6931b91b6b2fc5b308b1fb8ab3R77) doesn't seem to exist in [older python versions](https://github.com/python/cpython/blob/3.8/Lib/importlib/metadata.py#L41) and is an optional attribute in the [latest version](https://github.com/python/cpython/blob/3.12/Lib/importlib/metadata/__init__.py#L194).

File: opentelemetry-instrumentation/pyproject.toml
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/pyproject.toml:30]: https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2181#discussion_r1739214916

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (Arnatious) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:50]: I'm suggesting replacing all uses of `importlib.metadata` or `importlib_metadata` with `opentelemetry.util._importlib_metadata` and taking it as a transitive dependency from opentelemetry-api itself.

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (Arnatious) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:35]: I meant the _mapping attribute within _EntrypointDistFinder - it'll do exactly what you're already doing and only compute _mapping once

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (Arnatious) [inline opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:34]: Typo? `__int__` vs `__init__`

(Can probably just be deleted if https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2181/files#r1739126880 is accepted)

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:50]: Okay, my bad. It seems reasonable to use the alias.

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:35]: Okay yes, I agree

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (Rodrigo-Novas) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:34]: Yes it is a typo

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (xrmx) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:101]: If we are going to use importlib_metadata (via opentelemetry api util) we just need to be assured that it's there since 6+ no? Looking at some old pkg_resources code, dist was default None there too.

**Comment by serkan-ozal** [comment]
Switching from `pkg_resource` to `importlib.metadata` also helps us to reduce OTEL Python layer Lambda coldstart.

**Comment by bhaskarbanerjee** [comment]
When is this PR expected to be approved and merged?

**Comment by xrmx** [comment]
> When is this PR expected to be approved and merged?

When it's ready

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py
Comment (ocelotl) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/__init__.py:50]: Yes, please use `opentelemetry.util._importlib_metadata`. We defnitely *don't* want to use neither `importlib.metadata` nor `imporlib_metadata` anywhere else outside of `opentelemetry.util._importlib_metadata`. The implementation of `importlib.metadata` and `importlib_metadata` is plagued with API changes between Python versions and we want to contain all the issues this causes into one single place.

BTW, since there is a lot of interest in moving this forward, I added some commits on top of your PR in #2854. I cannot continue debugging the failing test cases right now but I encourage you to use the code in #2854, which removes usage of `importlib.metadata` in test requirements files and uses `opentelemetry.util._importlib_metadata` in source code.

Also, you'll notice that the `CORE_REPO_SHA` environment variable in #2854 is set to a particular commit, that commit should always be the latest commit of [this](https://github.com/open-telemetry/opentelemetry-python/pull/4177). Once that PR gets merged, the value of `CORE_REPO_SHA` should be set back to `main` in the `tox.ini` file and the `.github/workflows/*.yml` files.

**Comment by ocelotl** [comment]
> When is this PR expected to be approved and merged?

We are trying to get more eyes on this PR, we realize it is important for many people. The issue at hand here is not particularly easy to solve (kudos to @Rodrigo-Novas for the work so far!) since it deals with replacing an old library with a new one where there is not necessarily a 1:1 relationship between old library and new library components. To make matters even more complicated, the new library has been implemented very inconsistently, so its behavior for Python 3.x is not the same as for Python 3.y.

We are working on this and we'll merge as soon as we can :slightly_smiling_face:

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (xrmx) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:35]: @Arnatious isn't that missing the dist parameter though?

Comment (xrmx) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:35]: I've tried this (with the two for lines swapped) but it looks like the code is not used with our current tests, see 
https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2871/files#r1759148502

So the question is: do we really need this cache?

**Comment by Arnatious** [comment]
For anyone following - this is currently blocked by (at minimum) https://github.com/open-telemetry/opentelemetry-python/pull/4177, as well as a new release needed for https://github.com/open-telemetry/opentelemetry-python/pull/4181 .  Some of our assumptions in the previous discussions mentioned need to be re-addressed if we swap to relying on the stdlib in python <= 3.11, which will delay this a bit further.

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (Arnatious) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:101]: We're not going to be guaranteed 6+ if https://github.com/open-telemetry/opentelemetry-python/pull/4177 goes through - we'll only have 6 on python 3.12 and higher.

**Comment by xrmx** [comment]
> For anyone following - this is currently blocked by (at minimum) [open-telemetry/opentelemetry-python#4177](https://github.com/open-telemetry/opentelemetry-python/pull/4177), as well as a new release needed for [open-telemetry/opentelemetry-python#4181](https://github.com/open-telemetry/opentelemetry-python/pull/4181) . Some of our assumptions in the previous discussions mentioned need to be re-addressed if we swap to relying on the stdlib in python <= 3.11, which will delay this a bit further.

We're keeping importlib_metadata as dependency

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py
Comment (Arnatious) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:35]: Looks like according to coverage everything we test has a dist attribute, and therefore exits before the call to ._mapping

Comment (Arnatious) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:35]: Granted, not sure that's necessarily something _this_ pr needs to fix.

**Comment by multani** [comment]
@xrmx I think this should have been closed by #2871, right?

**Comment by xrmx** [comment]
Closing, since this has been included in #2871. Thanks everyone!

**Pull request closed without merging**
